### ISSUE TITLE
Proposal: Arguments enhancement.

### ISSUE THREAD

**Issue by kyranet** [dirigeants/klasa#161]
### Describe the issue

We can agree that **usage strings** are very cool, very easy to use, very clean and readable, compared to other systems implemented in other frameworks with are, for example, an array of objects that lately face the issue of `union` type, something easy to write thanks to the usage string. However, as easy as it is, the usage string has one weakness that we can't address in the current way, for that, I have multiple proposals.

---

### UsagePrompt

This is @MrJacz's proposal, an array typeof `Array<string | (msg: KlasaMessage) => string`. This is the easiest way to address this issue as it's another property for `CommandOptions`: `usagePrompt`. It's usage would come to be something similar to:

```javascript
{
    usage: '<battletag:string> [bronze|silver|gold|platinum]',
    usageDelim: ' ',
    usagePrompt: [
        'Give me a battletag formatted as username#0000',
        'Give me a rank name, can be bronze, silver, gold or platinum, pick the one that is next to your rank.'
    ]
}
```

However, it's lineal and to do so you may need to pass empty values or undefined in the middle of the array, to address that, it's possible to have a class do that:

```javascript
{
    usage: '<battletag:string> [bronze|silver|gold|platinum]',
    usageDelim: ' ',
    usagePrompt: new UsagePrompt()
        .addPrompt(1,
            'Give me a rank name, can be bronze, silver, gold or platinum, pick the one that is next to your rank.')
}
```

Where `0`, being `<battletag:string>`, would use the default message (`battletag is a required argument`).

---

However, it may be confusing, may it start with zero? Or maybe with one? To address this issue, we could make `CommandOptions.usage` accept more types: not just `string`.

### ArgumentUsage

```javascript
{
    usage: new ArgumentUsage()
        .addParameter('battletag', /(\w{4,12})#(\d{4,5})/, true,
            'Give me a valid BattleTag, should be formatted like someuser#0000')
        .addParameter('rank', 'bronze|silver|gold|platinum', false,
            'Give me a rank name, can be bronze, silver, gold or platinum, pick the one that is next to your rank.'),
    usageDelim: ' '
}
```

Where `ArgumentUsage` is a class that compiles directly to `ParsedUsage` and has a `.toString()` method to compile to `usageString` (used in the help command, for example). The reason this would compile to `ParseUsage` is so it's able to add the new properties to the possibles: `promptMessage` being either a string or callback `(msg: KlasaMessage) => string`.

The `Argument#addParameter` option is **very powerful** as it doesn't only add one more feature to Usage, which is prompt, but it also adds two more (nice) features: the possibility to have raw RegExp to use as regexp type (I'll be honest, I haven't figured the RegExp type since it got released, it's very complicated inside usageString, if this gets merged, the usage of the `RegExp` type in Klasa will be significantly easier) or even a **custom** type. Of course, this part of the concept can change, and name would be... perhaps `custom` or allow a fifth parameter which is parameter name for usageString. This can be reviewed later.

Furthermore, stringifying a RegExp for usageString is relatively easy, [`RegExp#toString()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/toString) does that.

This method would also support strings as **legacy** and would allow union type. Of course, this could use the current usageString parser, covering the downsides of the systems that are not based in quotedString.

> The design of this method is inspired on **PermissionLevels** to keep the design of the framework and its core pieces.

**EDIT**: Stringifying an `ArgumentUsage` instance would be as easy as mapping all the entries and write `argName:typeName/typeParam` then wrap with `<` and `>` or `[` and `]` depending on whether the value of the second parameter is true or false.

```typescript
class ArgumentUsage {

    public constructor();
    private _entries: object[];

    public addParameter(
        argName: string, 
        type: string | RegExp | (msg: KlasaMessage, arg: string) => string,
        required: boolean,
        promptMessage?: string | (msg: KlasaMessage) => string,
        typeName?: string): this;

    public toJSON(): object;
    public toString(): string;

}
```

---

**EDIT 2**: `ArgumentUsage` is **not** meant to replace the former `usageString`, but to complement as a more advanced tool that opens the gates for more creativity, implementing better debugging (IntelliSense, separation of the "parts" of each Tag, custom types, better RegExp readability...) and the possibilities to reach where `usageString` can't reach while also using the parser for the aforementioned to cover the issues this kind of system has in other frameworks (lack of union type as an example).

This is more than just an enhanced usage, this allows better debugging, better internationalization, custom messages/prompts (feature **highly requested over almost half a year**) and improves user friendly-ness towards users.

## Implementation

None.

## Further details

- [PermissionLevels](https://klasa.js.org/PermissionLevels.html)

**Comment by MrJacz** [comment]
I agree with this 100% all ideas look great as they all work but come with pro's and con's like everything does the `ArgumentUsage` way looks very nice and advanced but will this suit for beginners? not really IMO  while it is very advanced and would be powerful it could also be a hassle for beginners to understand but that being said it won't take long to learn how it works.

Now while I also like the idea of `UsagePrompt` I don't see how useful it could be considering if you're going to go full out on a prompting class for usage prompts I would just go with the `ArgumentUsage`  class idea which handles Usage as a whole excluding usageDelim as that has its differences as I see it  `ArgumentUsage`  seems the way to go its advanced and powerful unlike `UsagePrompt` which it would be plain and boring just handling prompts. yes, it could be advanced for handling prompts but why would it need to be? it doesn't it only got to handle prompts, having an ArumentUsage handling that seems better and fresher as it is not a copy of any other framework arg systems it's Klasa's own still using the same usage strings as we always were. not many changes there this could also make it so much easier and cleaner to do arguments as we could now pass a [**RegExp**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/) as the type. We also allow the use of `internationalization` also known as `i18n` in prompts this making `ArgumentUsage` the most advanced Usage handling in a framework because Union types any other array based arg systems in other frameworks failed to add union types as it is too hard but `ArgumentUsage` makes union types still usable like before, now like I was saying this makes `ArgumentUsage` one of the best Usage/Arg systems in Discord.js Frameworks due to the fact how Powerful and Advanced it can be.

### Suggestions / Ideas
 `paramName` changed to `id` or `key` as `paramName` seems quite long just for an "id" or "key" as it is just the usage name for prompting and usage strings in help. Of course, it can be used in other cases tho

Add min and max while this could still be handled in `type`.

**Comment by Soumil07** [comment]
@MrJacz in the server we have many people who get errors because of usages. I myself took a while to figure out how the UsageDelim thing worked.

My point is: a total beginner will not understand the UsageString itself at first. They'll take some time, and figure it out. And then UsagePrompt won't be too hard, seeing it's so similar to permLevel.

And I'm against the argument usage idea because it seems unnecessary. As Kyra said, the current usageString is very good because it can be written in a jiffy. So much longer with the argument usage class. If both usage and that class are accepted, it's fine. But it seems unnecessary for just prompts.

@kyranet I imagine the UsagePrompt won't be able to handle union types then? (Different prompts for different types.)

**Comment by MrJacz** [comment]
my point being, my array way is ugly and Ao will most likely dislike `something, undefined, next prompt here` and UsagePrompt doesn't seem quite useful as if anything I feel as if you're going to add that you should just make a class to handle Usage as a whole yet `usage` accept `ArgumentUsage` and a regular usage string but understanding how Usage and usage delim works takes time a couple of creating commands without help at least

**Comment by kyranet** [comment]
Replying to @MrJacz 

> everything does the `ArgumentUsage` way looks very nice and advanced but will this suit for beginners? not really IMO while it is very advanced and would be powerful it could also be a hassle for beginners to understand

If you use an IDE with [IntelliSense](https://code.visualstudio.com/docs/editor/intellisense), it'll reveal the type and name for each parameter, making it substantially easier. And `usageString` isn't that easy to debug, however.

> paramName changed to id or key as paramName seems quite long just for an "id" or "key" as it is just the usage name for prompting and usage strings in help.

In fact the correct and technical name is `ArgumentName` or `ArgumentID`, as Usage is understood as an *array* of arguments.

> Add min and max while this could still be handled in type.

Using the current parser, this is possible. When `type` is typeof string, Klasa would use [`Tag.parseMembers()`](https://github.com/dirigeants/klasa/blob/master/src/lib/usage/Tag.js#L38-L61) to parse the tags inside.

---

Replying to @Soumil07 

> And I'm against the argument usage idea because it seems unnecessary. As Kyra said, the current usageString is very good because it can be written in a jiffy. So much longer with the argument usage class. If both usage and that class are accepted, it's fine. But it seems unnecessary for just prompts.

In fact ArgumentUsage has several additions:

1. Customizable prompts and i18n messages.
2. Better RegExp debugging. Of course, RegExp will always be valid as its misstyping would cause in JS to fail parsing it. And you can use your IDE and the syntax highlight to debug and read it even better.
3. Custom types. You need an argument type **just** for a single command? Pass a function. Done.
4. (Optional, depends on future implementation) You have an array of 23 possibles (for example, one for each hero in Overwatch)? You could "mask" them and show `heroName`.

And both usageString and ArgumentUsage are accepted, yes.

> I imagine the UsagePrompt won't be able to handle union types then? (Different prompts for different types.)

That's not possible with `UsagePrompt` nor `ArgumentUsage`, in fact, it should keep one for each argument rather than one for each `Tag`.

**Comment by bdistin** [comment]
Closed in favor of #162
